### ISSUE TITLE
Patched method automatic unpatched?

### ISSUE THREAD

**Issue by pengweiqhca** [pardeike/Harmony#307]
**Describe the bug**
Patched method automatic unpatched?

**Expected behavior**
No exception

**Screenshots / Code**
``` C#
public static class Program
{
    private static async Task Main()
    {
        var patcher = new Harmony(nameof(SocketsHttpHandler));

        patcher.Patch(typeof(SocketsHttpHandler).Assembly.GetType("System.Net.Http.HttpConnectionPoolManager")!
                .GetMethod("GetConnectionKey", BindingFlags.Instance | BindingFlags.Static | BindingFlags.NonPublic),
            new HarmonyMethod(typeof(Program).GetMethod(nameof(GetConnectionKeyBefore), BindingFlags.Static | BindingFlags.NonPublic)),
            new HarmonyMethod(typeof(Program).GetMethod(nameof(GetConnectionKeyEnd), BindingFlags.Static | BindingFlags.NonPublic)));

        await Test();
        await Test();
        await Task.Delay(1000);
        await Test(); //Exception
    }

    private static async Task Test()
    {
        using var client = new HttpClient();

        using var request = new HttpRequestMessage(HttpMethod.Get, "https://iovip.qbox.me/test/FtPFFB2e_qh1isgop4jyxG68k4tE_w2880_h1800.jpg@200w_200h_100q.webp");

        request.Headers.Host = "image.tuhu.cn";

        using var response = await client.SendAsync(request);

        Console.WriteLine(response.StatusCode);
    }

    private static void GetConnectionKeyBefore(HttpRequestMessage request, out string __state)
    {
        __state = null;

        if (ParseHostNameFromHeader(request.Headers.Host) == request.RequestUri.IdnHost ||
            request.Headers.Host == null) return;

        __state = request.Headers.Host;
        request.Headers.Host = null;

        static string ParseHostNameFromHeader(string hostHeader)
        {
            if (hostHeader == null) return null;

            // See if we need to trim off a port.
            var colonPos = hostHeader.IndexOf(':');
            if (colonPos < 0) return hostHeader;

            var ipV6AddressEnd = hostHeader.IndexOf(']');
            if (ipV6AddressEnd == -1)
                return hostHeader.Substring(0, colonPos);

            colonPos = hostHeader.LastIndexOf(':');
            return colonPos > ipV6AddressEnd ? hostHeader.Substring(0, colonPos) : hostHeader;
        }
    }

    private static void GetConnectionKeyEnd(HttpRequestMessage request, string __state)
    {
        if (__state != null) request.Headers.Host = __state;
    }
```

**Runtime environment (please complete the following information):**
 - OS:  Windows 10 2004
 - .NET version: netcoreapp3.1
 - Harmony version 2.0.1
- Console

**Comment by pengweiqhca** [comment]
After `disable tiered JIT compilation`, it is work, I am waiting the next big thing.

**Comment by InCerryGit** [comment]
> After `disable tiered JIT compilation`, it is work, I am waiting the next big thing. 

 `tiered JIT compilation` will JIT your code again to achieve higher performance, so your patch hook is overwritten.
I'm sure we'll see this exciting feature soon,  It's about to be fixed in `MonoMod`. That's very exciting, See link for more details: 
https://github.com/MonoMod/MonoMod/pull/67

**Comment by pardeike** [comment]
Same issue and closed in favor of #337
